src/tactic/ring_exp.lean

```diff
@@ -0,0 +1,1430 @@
+  The `ex` type is structured according to a parameter `et : ex_type`.
+
+  Instead of allowing each operand to have arbitrary subexpressions as argument,
+  we have a single `ex_type` allowed for each argument.
+  This e.g. enforces that `+` associates to the right,
+  and `+` distributes over `*`.
+  -/
+inductive ex_type : Type
+| base : ex_type
+| sum : ex_type
+| prod : ex_type
+| exp : ex_type
+open ex_type
+
+/--
+  Each `ex` stores information for its normalization proof.
+
+  The `orig` expression is the expression that was passed to `eval`.
+
+  The `pretty` expression is the normalised form that the `ex` represents.
+  (I didn't call this something like `norm`, because there are already
+  too many things called `norm` in mathematics!)
+
+  The field `proof` contains an optional proof term of type `%%orig = %%pretty`.
+  The value `none` for the proof indicates that everything reduces to reflexivity.
+  (Which saves space in quite a lot of cases.)
+-/
+meta structure ex_info : Type :=
+(orig : expr) (pretty : expr) (proof : option expr)
+
+/--
+  The `ex` type is an abstract representation of an expression with `+`, `*` and `^`.
+  Those operators are mapped to the `sum`, `prod` and `exp` constructors respectively.
+
+  The `zero` constructor is the base case for `ex sum`, e.g. `1 + 2` is represented
+  by (something along the lines of) `sum 1 (sum 2 zero)`.
+
+  The `coeff` constructor is the base case for `ex prod`, and is used for numerals.
+  The code maintains the invariant that the coefficient is never `0`.
+
+  The `var` constructor is the base case for `ex exp`, and is used for atoms.
+
+  The `sum_b` constructor allows for addition in the base of an exponentiation;
+  it serves a similar purpose as the parentheses in `(a + b)^c`.
+  The code maintains the invariant that the argument to `sum_b` is not `zero`
+  or `sum _ zero`.
+
+  All of the constructors contain an `ex_info` field,
+  used to carry around (arguments to) proof terms.
+
+  While the `ex_type` parameter enforces some simplification invariants,
+  the following ones must be manually maintained at the risk of insufficient power:
+   - the argument to `coeff` must be nonzero (to ensure `0 = 0 * 1`)
+   - the argument to `sum_b` must be of the form `sum a (sum b bs)` (to ensure `(a + 0)^n = a^n`)
+   - normalisation proofs of subexpressions must be `refl ps.pretty`
+   - if we replace `sum` with `cons` and `zero` with `nil`, the resulting list is sorted
+     according to the `lt` relation defined further down; similarly for `prod` and `coeff`
+     (to ensure `a + b = b + a`).
+
+  The first two invariants could be encoded in a subtype of `ex`,
+  but aren't (yet) to spare some implementation burden.
+  The other invariants cannot be encoded because we need the `tactic` monad to check them.
+  (For example, the correct equality check of `expr` is `is_def_eq : expr → expr → tactic unit`.)
+-/
+meta inductive ex : ex_type → Type
+| zero  {} (info : ex_info) : ex sum
```

Comment:
```suggestion
| zero (info : ex_info) : ex sum
```
I don't think any of these should be necessary either. Or am I misunderstanding?

Reply:
I forgot to remove them after removing parameters to `ex`, so they are indeed unnecessary!